### PR TITLE
JS modules

### PR DESCRIPTION
**Pull request by letmaik**
- Add JS module support in CCF (new table, new proposals, module loader for quickjs)
- Change JS-based endpoints to be modules with a single (default) export
- Tests

### CODE REVIEW

File: src/node/modules.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#include <stdint.h>
+#include <string>
+#include <vector>
+
+namespace ccf
+{
+  struct Module
+  {
+    std::optional<std::string> js;
+    // std::optional<uint8_t> wasm;
```
Comment (eddyashton) [inline src/node/modules.h:16]: For now, I think we should just make this a required string member:

```suggestion
    std::string js;
```

File: src/node/modules.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  // Current limitation of the JSON macros: It is necessary to defined
+  // DECLARE_JSON_REQUIRED_FIELDS for Script even though there are no required
+  // fields. This raises some compiler warnings that are disabled locally.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-parameter"
+#pragma clang diagnostic ignored "-Wgnu-zero-variadic-macro-arguments"
+  DECLARE_JSON_REQUIRED_FIELDS(Module)
+#pragma clang diagnostic pop
+
+  DECLARE_JSON_OPTIONAL_FIELDS(Module, js)
```
Comment (eddyashton) [inline src/node/modules.h:37]: With above, I don't think this needs to optional:

```suggestion
  DECLARE_JSON_TYPE(Module)
  DECLARE_JSON_REQUIRED_FIELDS(Module, js)
```

File: tests/modules.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import infra.net
+import infra.e2e_args
+import suite.test_requirements as reqs
+import ccf.proposal_generator
+
+from loguru import logger as LOG
+
+MODULE_CONTENT = """
+export function foo() {
+    console.log("FOOO");
```
Comment (eddyashton) [inline tests/modules.py:19]: Remove or `console.log("Calling module function");

File: tests/modules.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        network.consortium.set_js_app(
+            remote_node=primary, app_script_path=f.name
+        )
+
+    with primary.client("user0") as c:
+        r = c.post(
+            "/app/test_module", {}
+        )
+        assert r.status_code == 200, r.status_code
+        assert r.body == "Hello world!"
```
Comment (eddyashton) [inline tests/modules.py:93]: Minor: You could avoid repeating this with something like

```
MODULE_RETURN = "Hello world!"
MODULE_CONTENT = f"""... return {MODULE_RETURN} ... """

...

assert r.body == MODULE_RETURN
```

File: 3rdparty/quickjs/quickjs.c [hunk 5]
```diff
@@ -34691,6 +34691,26 @@ int JS_SetModuleExportList(JSContext *ctx, JSModuleDef *m,
...
+    return JS_DupValue(ctx, m->export_entries[idx].u.local.var_ref->value);
+}
+
+JSAtom JS_GetModuleExportEntryName(JSContext *ctx, JSModuleDef *m, int idx)
+{
+    if (idx >= m->export_entries_count || idx < 0)
+        return JS_ATOM_NULL;
+    return JS_DupAtom(ctx, m->export_entries[idx].export_name);
+}
+
```
Comment (achamayou) [inline 3rdparty/quickjs/quickjs.c:34713]: I think we could just #include "../quickjs_patch/quickjs.c" here, and keep 3rdparty/quickjs_path with quickjs.c and quickjs.h (containing the extra signatures below).

It's not as generic as the patch, but it's a single, can't-go-out-sync copy of the changes, with a one-line, trivial-to-merge update to upstream. I'd prefer it, but I'm also happy to go with this is we have consensus.

File: src/node/rpc/member_frontend.h [hunk 6]
```diff
@@ -96,6 +104,28 @@ namespace ccf
...
 
+    void set_module(kv::Tx& tx, std::string name, Module module)
+    {
+      auto tx_modules = tx.get_view(network.modules);
+      tx_modules->put(name, module);
+    }
+
+    void remove_module(kv::Tx& tx, std::string name)
+    {
+      auto tx_modules = tx.get_view(network.modules);
```
Comment (jumaffre) [inline src/node/rpc/member_frontend.h:115]: What if the module does not exist?

File: src/node/rpc/member_frontend.h [hunk 7]
```diff
@@ -96,6 +104,28 @@ namespace ccf
...
+      tx_modules->remove(name);
+    }
+
+    Module get_module(kv::Tx& tx, std::string name)
```
Comment (jumaffre) [inline src/node/rpc/member_frontend.h:119]: Unused?

File: tests/modules.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    LOG.info("Member makes a module remove proposal")
+    proposal_body, _ = ccf.proposal_generator.remove_module("foo.js")
+    proposal = network.consortium.get_any_active_member().propose(primary, proposal_body)
+    network.consortium.vote_using_majority(primary, proposal)
+
+    with primary.client(f"member{network.consortium.get_any_active_member().member_id}") as c:
+        r = c.post(
+            "/gov/read", {"table": "ccf.modules", "key": "foo.js"}
+        )
+        assert r.status_code == 400, r.status_code
```
Comment (jumaffre) [inline tests/modules.py:64]: 400 -> `http.HTTPStatus.NOT_FOUND`

**Comment by achamayou** [comment]
Hm, I tried to update from master, it failed and had a retry button, which I pressed, and it merged!

**Comment by letmaik** [comment]
I'll do a follow-up PR.

File: src/node/rpc/member_frontend.h (hunk 6 above)
Comment (letmaik) [reply src/node/rpc/member_frontend.h:115]: Addressed in #1472.

File: src/node/rpc/member_frontend.h (hunk 7 above)
Comment (letmaik) [reply src/node/rpc/member_frontend.h:119]: Addressed in #1472.

File: tests/modules.py (hunk 8 above)
Comment (letmaik) [reply tests/modules.py:64]: Addressed in #1472.

File: 3rdparty/quickjs/quickjs.c (hunk 5 above)
Comment (letmaik) [reply 3rdparty/quickjs/quickjs.c:34713]: Addressed in #1472.

**Pull request merged**